packages/cli/src/PublicApi/v1/handlers/credentials/credentials.service.ts

```diff
@@ -0,0 +1,247 @@
+	properties: Partial<CredentialsEntity>,
+): Promise<CredentialsEntity> {
+	const newCredential = new CredentialsEntity();
+
+	Object.assign(newCredential, properties);
+
+	if (!newCredential.nodesAccess || newCredential.nodesAccess.length === 0) {
+		newCredential.nodesAccess = [
+			{
+				nodeType: `n8n-nodes-base.${properties.type?.toLowerCase() ?? 'unknown'}`,
+				date: new Date(),
+			},
+		];
+	} else {
+		// Add the added date for node access permissions
+		newCredential.nodesAccess.forEach((nodeAccess) => {
+			// eslint-disable-next-line no-param-reassign
+			nodeAccess.date = new Date();
+		});
+	}
+
+	return newCredential;
+}
+
+export async function saveCredential(
+	credential: CredentialsEntity,
+	user: User,
+	encryptedData: ICredentialsDb,
+): Promise<CredentialsEntity> {
+	const role = await Db.collections.Role.findOneOrFail({
+		name: 'owner',
+		scope: 'credential',
+	});
+
+	await externalHooks.run('credentials.create', [encryptedData]);
+
+	return Db.transaction(async (transactionManager) => {
+		const savedCredential = await transactionManager.save<CredentialsEntity>(credential);
+
+		savedCredential.data = credential.data;
+
+		const newSharedCredential = new SharedCredentials();
+
+		Object.assign(newSharedCredential, {
+			role,
+			user,
+			credentials: savedCredential,
+		});
+
+		await transactionManager.save<SharedCredentials>(newSharedCredential);
+
+		return savedCredential;
+	});
+}
+
+export async function removeCredential(credentials: CredentialsEntity): Promise<ICredentialsDb> {
+	await externalHooks.run('credentials.delete', [credentials.id]);
+	return Db.collections.Credentials.remove(credentials);
+}
+
+export async function encryptCredential(credential: CredentialsEntity): Promise<ICredentialsDb> {
+	const encryptionKey = await UserSettings.getEncryptionKey();
+
+	// Encrypt the data
+	const coreCredential = new Credentials(
+		{ id: null, name: credential.name },
+		credential.type,
+		credential.nodesAccess,
+	);
+
+	// @ts-ignore
+	coreCredential.setData(credential.data, encryptionKey);
+
+	return coreCredential.getDataToSave() as ICredentialsDb;
+}
+
+export function sanitizeCredentials(credentials: CredentialsEntity): Partial<CredentialsEntity>;
+export function sanitizeCredentials(
+	credentials: CredentialsEntity[],
+): Array<Partial<CredentialsEntity>>;
+
+export function sanitizeCredentials(
+	credentials: CredentialsEntity | CredentialsEntity[],
+): Partial<CredentialsEntity> | Array<Partial<CredentialsEntity>> {
+	const argIsArray = Array.isArray(credentials);
+	const credentialsList = argIsArray ? credentials : [credentials];
+
+	const sanitizedCredentials = credentialsList.map((credential) => {
+		// eslint-disable-next-line @typescript-eslint/no-unused-vars
+		const { data, nodesAccess, shared, ...rest } = credential;
+		return rest;
+	});
+
+	return argIsArray ? sanitizedCredentials : sanitizedCredentials[0];
+}
+
+export function toJsonSchema(properties: INodeProperties[]): IDataObject {
```

Comment:
could we add some comments for code documentation so we know what happens in which block?